### PR TITLE
Move dust collection hook to outside of account mutate

### LINKED ISSUE #7605: Possible reentrancy issues with hooks in mutate closure
Substrate have some hooks invocation inside `mutate` calls, this means there will be pending changes in stack which are not accessible by hooks. When hook attempting to read or modify the same values, a lost update will happen.

This could trigger some security issue if the implementer of the hooks did not know this constraint.

One example is the `DustRemoval` hook on pallet-balances

https://github.com/paritytech/substrate/blob/cff25fbc37c9fc564f8816eefdcd8dce15e1606b/frame/balances/src/lib.rs#L207

It is called in `post_mutation`

https://github.com/paritytech/substrate/blob/cff25fbc37c9fc564f8816eefdcd8dce15e1606b/frame/balances/src/lib.rs#L604-L611

which is called in `try_mutate_exists`

https://github.com/paritytech/substrate/blob/cff25fbc37c9fc564f8816eefdcd8dce15e1606b/frame/balances/src/lib.rs#L645-L654

So the `DustRemoval` cannot do any modification to the `maybe_account` or lost update will happen.

### PR DESCRIPTION
**Pull request by gavofyork**
Fixes #7605 (probably)

cc @xlc

### CODE REVIEW

File: frame/balances/src/lib.rs [hunk 1]
```diff
@@ -610,16 +610,17 @@ impl<T: Config<I>, I: Instance> Module<T, I> {
...
-	) -> Option<AccountData<T::Balance>> {
+	) -> (Option<AccountData<T::Balance>>, Option<NegativeImbalance<T, I>>) {
 		let total = new.total();
 		if total < T::ExistentialDeposit::get() {
-			if !total.is_zero() {
-				T::DustRemoval::on_unbalanced(NegativeImbalance::new(total));
+			if total.is_zero() {
+				(None, None)
+			} else {
 				Self::deposit_event(RawEvent::DustLost(who.clone(), total));
```
Comment (xlc) [inline frame/balances/src/lib.rs:619]: should we move emit DustLost event to after `DustRemoval::on_unbalanced`? previously event is emitted after handler, this make it before handler.

**Review by xlc — COMMENTED** [review]
I think this should fix my issue. But perhaps should incorporate some tests I had here to verify it actually fixes the issue https://github.com/laminar-protocol/substrate/commit/0fe33f078cd5242376e575136d0fd2d21d5627c9

File: frame/balances/src/lib.rs [hunk 2]
```diff
@@ -651,15 +652,20 @@ impl<T: Config<I>, I: Instance> Module<T, I> {
...
-				(maybe_endowed, result)
+				let maybe_account_maybe_dust = Self::post_mutation(who, account);
+				*maybe_account = maybe_account_maybe_dust.0;
+				(maybe_endowed, maybe_account_maybe_dust.1, result)
 			})
-		}).map(|(maybe_endowed, result)| {
+		});
+		result.map(|(maybe_endowed, maybe_dust, result)| {
+			if let Some(dust) = maybe_dust {
+				T::DustRemoval::on_unbalanced(dust);
```
Comment (thiolliere) [inline frame/balances/src/lib.rs:667]: in `fn transfer` and `fn repatriate_reserved`, `try_mutate_account` is called inside another `try_mutate_account`.

Thus `on_unbalanced` cannot change the balance of the `dest` of a transfer. and cannot change the balance of the `beneficiary` of a repatriate_reserved.

Maybe we need another PR, or change transfer and repatriate_reserved here to fix the issue completely

Comment (gavofyork) [reply frame/balances/src/lib.rs:667]: That's a separate issue though to do with nested `mutate_account`s, whereas this fixes the issue of a single `mutate_account` where the `on_unbalanced` happens to touch the account being mutated.

I fear with nested `mutate_account`s, we'll always see issues like this and we'll probably need to alter the pattern to fix it properly. Something like:

```
{
  let u32_wrapper_value: SomeSpecialType<u32> = U32StorageItem::get_mutable();
  *u32_wrapper_value += 1;
  // calling `.flush()` would flush changes into the storage item
  u32_wrapper_value.flush();
  // other stuff could be written in afterwards
  *u32_wrapper_value = 0;
}
// After the `SomeSpecialType` is dropped, any pending changes are written, like with `.flush()`
```

**Review by thiolliere — APPROVED** [review]
We can fix the nested `mutate_account` of pallet-balances in another PR,
This PR looks good to me.
I have no opinon on the order of DustLost event

File: frame/balances/src/lib.rs (hunk 2 above)
Comment (thiolliere) [reply frame/balances/src/lib.rs:667]: yes I agree, maybe nested `mutate_account` calls of pallet-balances can just be rewritten, like with https://github.com/paritytech/substrate/pull/7783
Or introducing another abstraction

File: frame/balances/src/lib.rs (hunk 1 above)
Comment (gavofyork) [reply frame/balances/src/lib.rs]: Events are going to change order with the proper fix.

**Comment by gavofyork** [comment]
Fixed that additional case with a deferred dust cleanup type. (@thiolliere)

**Comment by gavofyork** [comment]
@xlc would be happy to accept a PR placing additional tests in here.

File: frame/balances/src/lib.rs [hunk 3]
```diff
@@ -1322,18 +1368,28 @@ impl<T: Config<I>, I: 'static> ReservableCurrency<T::AccountId> for Pallet<T, I>
...
-					Status::Free => to_account.free = to_account.free.checked_add(&actual).ok_or(Error::<T, I>::Overflow)?,
-					Status::Reserved => to_account.reserved = to_account.reserved.checked_add(&actual).ok_or(Error::<T, I>::Overflow)?,
-				}
-				from_account.reserved -= actual;
-				Ok(actual)
-			})
-		})?;
+		let ((actual, _maybe_one_dust), _maybe_other_dust) = Self::try_mutate_account_with_dust(
+			beneficiary,
+			|to_account, is_new|-> Result<(Self::Balance, DustCleaner<T, I>), DispatchError> {
```
Comment (thiolliere) [inline frame/balances/src/lib.rs:1373]:
```suggestion
			|to_account, is_new| -> Result<(Self::Balance, DustCleaner<T, I>), DispatchError> {
```

**Review by thiolliere — APPROVED** [review]
Great, LGTM

**Comment by xlc** [comment]
> @xlc would be happy to accept a PR placing additional tests in here.

@shamb0 do you want to do that?

**Comment by shamb0** [comment]
Thanks @xlc , I am interested to pick up the task of verification.

**Comment by gavofyork** [comment]
@shawntabrizi could i get a 👍 please?

**Comment by gavofyork** [comment]
@kianenigma could i get a review on this please?

**Review by kianenigma — APPROVED** [review]
LGTM

**Comment by gavofyork** [comment]
bot merge

**Pull request merged**